File: 2025/day-1/part-two.ts
const FULL_CIRCLE = 100;

function normalize(position: number): number {
  return ((position % FULL_CIRCLE) + FULL_CIRCLE) % FULL_CIRCLE;
}

export function partTwo(input: Array<string>) {
  let position = 50;
  let result = 0;

  for (const turn of input) {
    const direction = turn[0];
    const value = Number(turn.slice(1));

    let firstZeroAt: number;

    if (direction === "L") {
      firstZeroAt = position === 0 ? FULL_CIRCLE : position;
    } else {
      firstZeroAt = position === 0 ? FULL_CIRCLE : FULL_CIRCLE - position;
    }

    let spins = 0;
    if (value >= firstZeroAt) {
      spins = 1 + Math.floor((value - firstZeroAt) / FULL_CIRCLE);
    }

    result += spins;

    if (direction === "L") {
      position -= value;
    } else {
      position += value;
    }

    position = normalize(position);
  }

  return result;
}

File: 2025-rust/day-1/src/part_two.rs
const FULL_CIRCLE: i32 = 100;

pub fn main(input: &Vec<&str>) -> i32 {
    let mut position = 50;
    let mut result = 0;

    for turn in input {
        let direction = turn.chars().next().unwrap();
        let value: i32 = turn[1..].parse().unwrap();

        let first_zero_at = match (direction, position) {
            ('L', 0) => FULL_CIRCLE,
            ('R', 0) => FULL_CIRCLE,
            ('L', p) => p,
            ('R', p) => FULL_CIRCLE - p,
            _ => 0,
        };

        result += if value.ge(&first_zero_at) {
            1 + (value - first_zero_at) / FULL_CIRCLE
        } else {
            0
        };

        position += match direction {
            'L' => -value,
            'R' => value,
            _ => 0,
        };

        position = position.rem_euclid(FULL_CIRCLE);
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_test() {
        let input = vec![
            "L68", "L30", "R48", "L5", "R60", "L55", "L1", "L99", "R14", "L82",
        ];

        let result = main(&input);
        assert_eq!(result, 6);
    }
}
